notes/lecture5.tex

```diff
@@ -0,0 +1,141 @@
+\end{equation}
+\\We use induction in order to prove $f(x_t)-f(x^*) \leq \frac{2\beta D^2}{t+2}$  based on \equationref{convergenceInequal} above.\\
+First step: Base Case $t=0$ \\
+Since $ f(x_{t+1})-f(x^*) \leq (1-\eta_t)(f(x_t)-f(x^*)) + \frac{\eta_t^2\beta D ^2}{2}$, when t=0, $\eta_t =\frac{2}{0+2}=1, then $
+\begin{align*} f(x_1)-f(x^*) &\leq (1-\eta_t)(f(x_t)-f(x^*)) + \frac{\beta}{2}\| x_1-x^* \|^2 \\
+&= (1-1)(f(x_t)-f(x^*)) + \frac{\beta}{2}\| x_1-x^* \|^2 \\
+&\leq \frac{\beta D^2}{2}\\
+&\leq \frac{2\beta D^2}{3}
+\end{align*}
+Thus, the induction hypothesis holds for our base case. \\
+(Note that we cannot directly use $f(x_{t+1})-f(x^*) \leq (1-\eta_t)(f(x_t)-f(x^*)) + \frac{\eta_t^2\beta D ^2}{2}$ directly to derivate the base case out. Because $(1-\eta_t)(f(x_t)-f(x^*))$ is greater than 0 and cannot be directly proved to be bounded.)\\
+Second step: We assume that $f(x_t)-f(x^*) \leq \frac{2\beta D^2}{t+2}$ holds $\forall t$ and can show that the induction hyposthesis holds in general.\\
+General case: given $f(x_t)$ holds for our destinated inequality, we'd like to show it also holds for $f(x_{t+1})$. \\
+Using \equationref{convergenceInequal},
+\begin{align*} f(x_{t+1})-f(x) &\leq (1- \frac{2}{t+2}) (f(x_t)-f(x^*))+ \frac{4}{2(t+2)} \beta D^2 \\
+& \leq (1- \frac{2}{t+2}) (\frac{2\beta D^2}{t+2})+ \frac{4}{2(t+2)} \beta D^2 \\
+& = \beta D^2 (\frac{2t}{(t+2)^2} +\frac{2}{(t+2)^2}) \\
+& = 2\beta D^2 \frac{t+1}{(t+2)^2} \\
+& = 2\beta D^2 (\frac{t+1}{t+2}) (\frac{1}{t+2}) \\
+& \leq 2\beta D^2 (\frac{t+2}{t+3}) (\frac{1}{t+2}) \\
+& = 2\beta D^2 \frac{1}{t+3}
+\end{align*}
+Thus, the inequality also holds for the t+1 case.
+
+\end{proof}
+
+\subsection{Examples}
+The code for the following examples can be found \href{https://ee227c.github.io/code/lecture4.html}{here}.
+
+\subsubsection{Nuclear norm projection}
+The \textit{nuclear norm} (sometimes called \textit{Schatten $1$-norm} or \textit{trace norm}) of a matrix $A$, denoted $\|A\|_*$, is defined as the sum of its singular values
+\[
+\|A\|_* = \sum_i \sigma_i(A)\,.
+\]
+The norm can be computed from the singular value decomposition of $A$.
+We denote the unit ball of the nuclear norm by 
+\[
+B_*^{m\times n}=\{A\in\mathbb{R}^{m\times n} \mid \|A\|_*\le 1\}.
+\]
+How can we project a matrix $A$ onto $B_*$? Formally, we want to solve
+\[
+\min_{X\in B_*}\|A-X\|_F^2
+\]
+Due to the rotational invariance of the Frobenius norm, the solution is obtained by projecting the singular values onto the unit simplex. This operation corresponds to shifting all singular values by the same parameter $\theta$ and clipping values at $0$ so that the sum of the shifted and clipped values is equal to $1$. This algorithm can be found from \cite{Duchi2008}.
+
+\subsubsection{Low-rank matrix completion}
+Suppose we have a partially observable matrix $Y$ and we would like to 
+find its completion form projected in a nuclear norm ball. Formally we have the objective function
```

Comment:
Please introduce what `O` is and how it relates to the observed entries. I mainly used the componentwise product with O in the code. For the math, we used $P_{O}$ which is a linear projection onto a subset of the coordinates specified by O.

Reply:
Generally, this is not an adequate introduction of the problem. Please give sufficient description for the reader to be able to follow.